### PR TITLE
Xshard pandas on Spark

### PR DESCRIPTION
**Pull request by dding3**
Add xshard panda on spark implementation and example

### CODE REVIEW

**Comment by dding3** [comment]
This PR depends on https://github.com/intel-analytics/analytics-zoo/pull/2115 .

File: pyzoo/zoo/xshard/shard.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            self.partition_id_list = list(chunk(self.object_ids, num_partitions))
+        else:
+            self.partition_id_list = list(chunk(self.collect_ids(), num_partitions))
+        return self
+
+    def get_partitions(self):
+        return self.partition_id_list
+
+
+class ScDataShards(DataShards):
```
Comment (jason-dai) [inline pyzoo/zoo/xshard/shard.py:64]: SparkDataShards

File: pyzoo/zoo/xshard/pandas/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+def read_csv_ray(context, file_path):
+    return read_file_ray(context, file_path, "csv")
+
+
+def read_json_ray(context, file_path):
+    return read_file_ray(context, file_path, "json")
+
+
+def read_csv_sc(context, file_path):
```
Comment (jason-dai) [inline pyzoo/zoo/xshard/pandas/preprocessing.py:80]: read_csv_spark

File: pyzoo/zoo/xshard/pandas/preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    return read_file_sc(context, file_path, "csv")
+
+
+def read_json_sc(context, file_path):
```
Comment (jason-dai) [inline pyzoo/zoo/xshard/pandas/preprocessing.py:84]: read_json_spark

File: pyzoo/zoo/xshard/pandas/preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+                           in list(chunk(file_paths, num_partitions)) if partition]
+    shards = [RayPandasShard.remote() for i in range(len(file_partition_list))]
+    [shard.read_file_partitions.remote(file_partition_list[i], file_type)
+     for i, shard in enumerate(shards)]
+    # shard_partitions = [[shard] for shard in shards]
+    data_shards = RayDataShards(shards)
+    return data_shards
+
+
+def read_file_sc(context, file_path, file_type):
```
Comment (jason-dai) [inline pyzoo/zoo/xshard/pandas/preprocessing.py:126]: read_file_spark

File: pyzoo/zoo/xshard/pandas/preprocessing.py [hunk 5]
```diff
@@ -14,15 +14,41 @@
...
 #
 import os
-
 import ray
 from pyspark.context import SparkContext
 
+from bigdl.util.common import get_node_and_core_number
+
+from zoo.common import get_file_list
 from zoo.ray.util.raycontext import RayContext
```
Comment (hkvision) [inline pyzoo/zoo/xshard/pandas/preprocessing.py:23]: need to change to `from zoo.ray import RayContext` to resolve conflicts... Sorry about that

Comment (dding3) [reply pyzoo/zoo/xshard/pandas/preprocessing.py]: Updated, thx~~

File: pyzoo/zoo/xshard/pandas/preprocessing.py [hunk 6]
```diff
@@ -123,12 +131,81 @@ def read_file_ray(context, file_path, file_type):
...
+def read_file_spark(context, file_path, file_type):
+    file_url_splits = file_path.split("://")
+    prefix = file_url_splits[0]
+    node_num, core_num = get_node_and_core_number()
+
+    if prefix == "s3":
+        data_paths = list_s3_file(file_url_splits, file_type)
+    else:
+        data_paths = get_file_list(file_path)
+    rdd = context.parallelize(data_paths, node_num * 20)
```
Comment (jason-dai) [inline pyzoo/zoo/xshard/pandas/preprocessing.py:143]: why `* 20`? should get `core_num`?

**Comment by jason-dai** [comment]
Is it ready for merge @dding3 @jenniew ?

**Comment by dding3** [comment]
It's waiting for performance test, but I can think it can be merged if needed

**Comment by dding3** [comment]
jekins passed. http://10.239.47.210:18888/job/ZOO-PR-Validation/3288/

**Comment by jason-dai** [comment]
> Update in a new PR #2292 . Yes, some partitions may be empty. It's a recommend partition size, and spark will adjust it.

If some partitions are empty, will it throw error? Shall we check it in the code?

**Pull request merged**